ambry-store/src/main/java/com.github.ambry.store/StorageManager.java

```diff
@@ -252,6 +257,27 @@ public boolean shutdownBlobStore(PartitionId id) {
     return failToUpdateStores;
   }
 
+  /**
+   * Update the hashmap of I/O error count per store with given {@link PartitionId} {@code id}.
+   * @param id the {@link PartitionId} of the {@link Store} whose count of I/O error should be updated.
+   * @param isStoreIOError whether to increase the error count ({@code true}) or reset it.
+   */
+  public void updateIOErrorCountPerStore(PartitionId id, boolean isStoreIOError) {
+    if (isStoreIOError) {
+      errorCountPerStore.compute(id, (k, v) -> {
+        v = (v == null) ? 0 : v;
+        if (v >= storeConfig.storeIOErrorCountToTriggerShutdown) {
+          shutdownBlobStore(id);
```

Comment:
Need to think on this a little bit. This can happen in the context of a request thread, is that correct? I'm not sure if that is a good idea. Another implementation would be add a job on the scheduler that regularly checks for any stores that have > some amount of errors and shuts them down asynchronously (or just schedule the job to shutdown right here instead of doing it in this thread).